examples/list.c

```diff
@@ -64,35 +62,34 @@ void step_map_inc(ssm_act_t *act) {
 
   switch (act->pc) {
   case 0:
-    switch (ssm_to_obj(cont->l)->mm.tag) {
+    switch (ssm_tag(cont->l)) {
     case Nil:
       goto match_Nil_0;
     case Cons:
       goto match_Cons_0;
     }
     SSM_ASSERT(0);
   match_Nil_0:
-    ssm_drop(&ssm_to_obj(cont->l)->mm);
-    *cont->__ret = ssm_from_obj(ssm_new(List_size, Nil));
+    *cont->__ret = ssm_marshal(Nil);
     break;
   match_Cons_0:;
-    ssm_value_t __tmp0 = ssm_to_obj(cont->l)->payload[0];
-    ssm_value_t __tmp1 = ssm_to_obj(cont->l)->payload[1];
-    ssm_dup(&ssm_to_obj(__tmp1)->mm);
+    ssm_value_t __i = ssm_to_obj(cont->l)[0];
+    ssm_value_t __l = ssm_to_obj(cont->l)[1];
+    if (ssm_on_heap(__l))
```

Comment:
Wart: Because the "heapness" of a value is no longer type-dependent, we need to check each value we want to drop or dup. Though perhaps not entirely a wart, since here we're trading off a little bit of runtime efficiency for massive memory efficiency.

Reply:
I "fixed" this by hiding the check inside `ssm_{dup,drop,reuse}`, which are now macros. Under the hood, they call `ssm_{dup,drop,reuse}_unsafe`, which can be called instead when the caller is certain that the value points to the heap.

Reply:
This should also make it easier to generate code.